[DPE-5355] Deployment docs for AKS + EKS (#146)
canonical/kafka-k8s-operator: docs/how-to/h-deploy-eks.md

# How to deploy on EKS

[Amazon Elastic Kubernetes Service](https://aws.amazon.com/eks/) (EKS) is a popular, fully automated Kubernetes service. To access the EKS Web interface, go to [console.aws.amazon.com/eks/home](https://console.aws.amazon.com/eks/home).

## Summary
* [Install EKS and Juju tooling](#install-eks-juju)
* [Create a new EKS cluster](#create-eks-cluster)
* [Bootstrap Juju on EKS](#boostrap-juju)
* [Deploy charms](#deploy-charms)
* [Display deployment information](#display-information)
* [Clean up](#clean-up)

---

## Install EKS and Juju tooling

> Install [Juju](https://juju.is/docs/juju/install-juju) and the [`kubectl` CLI tools](https://kubernetes.io/docs/tasks/tools/) via snap:

Reviewer: explain that `kubectl` is for managing Kubernetes clusters

      minSize: 3
      maxSize: 5
      iam:
        attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
        - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
        - arn:aws:iam::aws:policy/AmazonS3FullAccess
      instancesDistribution:
        maxPrice: 0.15
        instanceTypes: ["m5.xlarge", "m5.2xlarge"] # At least two instance types should be specified
        onDemandBaseCapacity: 0
        onDemandPercentageAboveBaseCapacity: 50
        spotInstancePools: 2
EOF
```

Bootstrap EKS cluster with the following command:

```shell
eksctl create cluster -f cluster.yaml
```

Sample output:

```shell
...
2023-10-12 11:13:58 [ℹ]  using region eu-west-3
2023-10-12 11:13:59 [ℹ]  using Kubernetes version 1.27
...
2023-10-12 11:40:00 [✔]  EKS cluster "eks-taurus-27506" in "eu-west-3" region is ready
```

## Bootstrap Juju on EKS

Add Juju K8s clouds:

```shell
juju add-k8s $JUJU_NAME
```

Bootstrap Juju controller:

```shell
juju bootstrap $JUJU_NAME
```

## Deploy Charms

Create a new Juju model, if needed:

```shell

> juju add-model <MODEL_NAME>

Reviewer: provide a working example. This command as typed will fail

Author: it is expected, as it is a how-to and not a tutorial.

```

> > (Optional) Increase the debug level if you are troubleshooting charms:

Reviewer: this is superfluous

Author: I have revised this to be consistent with VM.

```

And integrate it with the Kafka application:

```shell
juju integrate kafka-k8s admin
```

For more information on Data Integrator and how to use it, please refer to the [how-to manage applications](/t/charmed-kafka-how-to-manage-app/10285) user guide.

## Display deployment information

Display information about the current deployments with the following commands:

```shell
~$ kubectl cluster-info 
Kubernetes control plane is running at https://AAAAAAAAAAAAAAAAAAAAAAA.gr7.eu-west-3.eks.amazonaws.com
CoreDNS is running at https://AAAAAAAAAAAAAAAAAAAAAAA.gr7.eu-west-3.eks.amazonaws.com/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

~$ eksctl get cluster -A
NAME            REGION      EKSCTL CREATED
eks-marc-9587	eu-west-3	True

~$ kubectl get node
NAME                                           STATUS   ROLES    AGE     VERSION
ip-192-168-1-168.eu-west-3.compute.internal    Ready    <none>   5d22h   v1.27.16-eks-a737599
ip-192-168-45-234.eu-west-3.compute.internal   Ready    <none>   3h25m   v1.27.16-eks-a737599
ip-192-168-85-225.eu-west-3.compute.internal   Ready    <none>   5d22h   v1.27.16-eks-a737599
```

## Clean up

[note type="caution"]

> Always clean EKS resources that are no longer necessary -  they could be costly!

Reviewer: "Always **terminate**..."

[/note]

To clean the EKS cluster, resources and juju cloud, run the following commands:

```shell
juju destroy-controller $JUJU_NAME --yes --destroy-all-models --destroy-storage --force
juju remove-cloud $JUJU_NAME
```

List all services and then delete those that have an associated EXTERNAL-IP value (e.g. load balancers):

```shell
kubectl get svc --all-namespaces
kubectl delete svc <service-name> 
```

Next, delete the EKS cluster  (source: [Deleting an Amazon EKS cluster]((https://docs.aws.amazon.com/eks/latest/userguide/delete-cluster.html) )) 

```shell
eksctl get cluster -A
eksctl delete cluster <cluster_name> --region eu-west-3 --force --disable-nodegroup-eviction
```

> Finally, remove AWS CLI user credentials (to avoid forgetting and leaking):

Reviewer: "leaking **credentials**"

# How to deploy on EKS

[Amazon Elastic Kubernetes Service](https://aws.amazon.com/eks/) (EKS) is a popular, fully automated Kubernetes service. To access the EKS Web interface, go to [console.aws.amazon.com/eks/home](https://console.aws.amazon.com/eks/home).

> ## Summary

Reviewer 2: ```suggestion
## Summary

```

* [Install EKS and Juju tooling](#install-eks-juju)
* [Create a new EKS cluster](#create-eks-cluster)
* [Bootstrap Juju on EKS](#boostrap-juju)
* [Deploy charms](#deploy-charms)
* [Display deployment information](#display-information)
* [Clean up](#clean-up)

---

## Install EKS and Juju tooling

Install [Juju](https://juju.is/docs/juju/install-juju) and the [`kubectl` CLI tools](https://kubernetes.io/docs/tasks/tools/) (that will be used for managing the Kubernetes cluster) via snap:

```shell
sudo snap install juju --channel 3.5/stable
sudo snap install kubectl --classic
```

> 
> 

Reviewer 2: ```suggestion
```

> Follow the installation guides for:

Reviewer 2: ```suggestion
Follow the installation guides for:

```

* [eksctl](https://eksctl.io/installation/) - the Amazon EKS CLI
* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) - the Amazon Web Services CLI

To check they are all correctly installed, you can run the commands below.

```shell
juju version
```

[details="Sample output:"]
```shell
3.5.2-genericlinux-amd64
```
[/details]

```shell
kubectl version --client
``` 

[details="Sample output:"]
```shell
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
```
[/details]

```shell
eksctl info
```

[details="Sample output:"]
```shell
eksctl version: 0.159.0
kubectl version: v1.28.2
```
[/details]

```shell
aws --version
```

[details="Sample output:"]
```shell
aws-cli/2.13.25 Python/3.11.5 Linux/6.2.0-33-generic exe/x86_64.ubuntu.23 prompt/off
```
[/details]

### Authenticate

Create an IAM account (or use legacy access keys) and login to AWS:

```shell

> > aws configure

Reviewer 2: Not a biggie, but can we fix the usage of the prompt for combining input and output here, since we do it everywhere else? As far as I can see this is the last place to fix it (remove the `>` and split output from input with proper introduction.

Author: I actually purposely didn't want to do that since it is not JUST output. The lines below should also provide guidance on how to feed the inputs (although I know it is obvious), and we actually use placeholder instead of just taking the output of a screen

AWS Access Key ID [None]: SECRET_ACCESS_KEY_ID
AWS Secret Access Key [None]: SECRET_ACCESS_KEY_VALUE
Default region name [None]: eu-west-3
Default output format [None]:
```

Verify that the CLI tool is correctly authenticating

```shell
aws sts get-caller-identity
```

[details="Sample output:"]
```yaml
{
    "UserId": "1234567890",
    "Account": "1234567890",
    "Arn": "arn:aws:iam::1234567890:root"
}
```
[/details]

## Create a new EKS cluster

> Export the deployment name for further use:

Reviewer 2: ```suggestion
Export the deployment name for further use:

```

        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
        - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
        - arn:aws:iam::aws:policy/AmazonS3FullAccess
      instancesDistribution:
        maxPrice: 0.15
        instanceTypes: ["m5.xlarge", "m5.2xlarge"] # At least two instance types should be specified
        onDemandBaseCapacity: 0
        onDemandPercentageAboveBaseCapacity: 50
        spotInstancePools: 2
```
[/details]

Bootstrap EKS cluster with the following command:

```shell
eksctl create cluster -f cluster.yaml
```

[details="Sample `cluster.yaml`:"]
```shell
...
2023-10-12 11:13:58 [ℹ]  using region eu-west-3
2023-10-12 11:13:59 [ℹ]  using Kubernetes version 1.27
...
2023-10-12 11:40:00 [✔]  EKS cluster "eks-taurus-27506" in "eu-west-3" region is ready
```
[/details]

## Bootstrap Juju on EKS

Add Juju K8s clouds:

```shell
juju add-k8s $JUJU_NAME
```

Bootstrap Juju controller:

```shell
juju bootstrap $JUJU_NAME
```

## Deploy Charms

Create a new Juju model, if needed:

```shell
juju add-model <MODEL_NAME>
```

> [note type="caution"]

Reviewer 2: Are you sure we need a caution here? I think Information/note would look better in this case. Just a suggestion.
```suggestion
[note]
```

> (Optional) Increase the debug level if you are troubleshooting charms:

Reviewer 2: ```suggestion
(Optional) Increase the debug level if you are troubleshooting charms:

```

```
[/details]

```shell
eksctl get cluster -A
```

[details="Sample output:"]
```shell
NAME            REGION      EKSCTL CREATED
eks-marc-9587	eu-west-3	True
```
[/details]

```shell
kubectl get node
```

[details="Sample output:"]
```shell
NAME                                           STATUS   ROLES    AGE     VERSION
ip-192-168-1-168.eu-west-3.compute.internal    Ready    <none>   5d22h   v1.27.16-eks-a737599
ip-192-168-45-234.eu-west-3.compute.internal   Ready    <none>   3h25m   v1.27.16-eks-a737599
ip-192-168-85-225.eu-west-3.compute.internal   Ready    <none>   5d22h   v1.27.16-eks-a737599
```
[/details]

## Clean up

[note type="caution"]
Always clean EKS resources that are no longer necessary -  they could be costly!
[/note]

To clean the EKS cluster, resources and juju cloud, run the following commands:

```shell
juju destroy-controller $JUJU_NAME --yes --destroy-all-models --destroy-storage --force
juju remove-cloud $JUJU_NAME
```

List all services and then delete those that have an associated EXTERNAL-IP value (e.g. load balancers):

```shell
kubectl get svc --all-namespaces
kubectl delete svc <service-name> 
```

> Next, delete the EKS cluster  (source: [Deleting an Amazon EKS cluster]((https://docs.aws.amazon.com/eks/latest/userguide/delete-cluster.html) )) 

Reviewer 2: ```suggestion
Next, delete the EKS cluster (As described on the [Deleting an Amazon EKS cluster](https://docs.aws.amazon.com/eks/latest/userguide/delete-cluster.html) page):
```
